extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/health/MongoHealthCheck.java

```diff
@@ -17,40 +17,69 @@
-        for (Bean<?> bean : beans) {
-            if (bean.getName() == null) {
-                // this is the default mongo client: retrieve it by type
-                MongoClient defaultClient = Arc.container().instance(MongoClient.class).get();
-                clients.put(DEFAULT_CLIENT, defaultClient);
-            } else {
-                MongoClient client = (MongoClient) Arc.container().instance(bean.getName()).get();
-                clients.put(bean.getName(), client);
+        for (InstanceHandle<MongoClient> handle : Arc.container().select(MongoClient.class, Any.Literal.INSTANCE).handles()) {
+            String clientName = getMongoClientName(handle.getBean());
+            clients.put(clientName == null ? CLIENT_DEFAULT : clientName, handle.get());
+        }
+        // reactive clients
+        for (InstanceHandle<ReactiveMongoClient> handle : Arc.container()
+                .select(ReactiveMongoClient.class, Any.Literal.INSTANCE).handles()) {
+            String clientName = getMongoClientName(handle.getBean());
+            reactiveClients.put(clientName == null ? CLIENT_DEFAULT_REACTIVE : clientName, handle.get());
+        }
+    }
+
+    /**
+     * Get mongoClient name if defined.
+     *
+     * @param bean
+     * @return mongoClient name or null if not defined
+     * @see MongoClientName
+     */
+    private String getMongoClientName(Bean bean) {
+        for (Object qualifier : bean.getQualifiers()) {
+            if (qualifier instanceof MongoClientName) {
+                return ((MongoClientName) qualifier).value();
             }
         }
+        return null;
     }
 
     @Override
     public HealthCheckResponse call() {
         HealthCheckResponseBuilder builder = HealthCheckResponse.named("MongoDB connection health check").up();
+        Document command = new Document("ping", 1);
         for (Map.Entry<String, MongoClient> client : clients.entrySet()) {
-            boolean isDefault = DEFAULT_CLIENT.equals(client.getKey());
-            MongoClient mongoClient = client.getValue();
             try {
-                Document document = mongoClient.getDatabase("admin").runCommand(new Document("ping", 1));
-                String mongoClientName = isDefault ? "default" : client.getKey();
-                builder.up().withData(mongoClientName, document.toJson());
+                Document document = client.getValue().getDatabase("admin").runCommand(command);
+                builder.up().withData(client.getKey(), document.toJson());
+            } catch (Exception e) {
+                return builder.down().withData("reason", "client [" + client.getKey() + "]: " + e.getMessage()).build();
+            }
+        }
+        for (Map.Entry<String, ReactiveMongoClient> client : reactiveClients.entrySet()) {
+            try {
+                Document document = client.getValue().getDatabase("admin").runCommand(command).await().indefinitely();
```

Comment:
Instead of blocking indefinitely, maybe we can introduce a timeout of `10s` with something like `.await()..atMost(Duration.ofSeconds(10));`.

Reply:
I was thinking about it as well but the reality is that the timeout is controlled by mongodb driver and its timeouts so once mongodb driver timeout for such operation it throws the exception.

Reply:
OK then I'll merge it as is.